File: frontend-client/src/api/viewing.ts
import { apiFetch } from './client'

export interface TitleInfo {
  title: string
  poster_url: string | null
  landscape_url: string | null
  title_type: 'movie' | 'series'
  age_rating: string | null
  episode_title?: string | null
  season_number?: number | null
  episode_number?: number | null
}

export interface NextEpisodeInfo {
  episode_id: string
  season_number: number
  episode_number: number
  episode_title: string
}

export interface ContinueWatchingItem {
  id: string
  content_type: 'movie' | 'episode'
  content_id: string
  position_seconds: number
  duration_seconds: number
  progress_percent: number
  completed: boolean
  dismissed_at: string | null
  updated_at: string
  resumption_score: number | null
  title_info: TitleInfo
  next_episode: NextEpisodeInfo | null
}

export interface BookmarkResponse {
  id: string
  content_type: string
  content_id: string
  position_seconds: number
  duration_seconds: number
  completed: boolean
  dismissed_at: string | null
  updated_at: string
}

export interface WatchlistItem {
  title_id: string
  title_info: {
    id: string
    title: string
    poster_url: string
    landscape_url: string | null
    title_type: 'movie' | 'series'
    release_year: number
    age_rating: string
  }
}

export interface BookmarkPayload {
  content_type: 'movie' | 'episode'
  content_id: string
  position_seconds: number
  duration_seconds: number
}

export interface RatingPayload {
  title_id: string
  rating: 1 | -1
}

export function getContinueWatching(
  profileId: string,
  deviceType?: string,
  hourOfDay?: number,
): Promise<ContinueWatchingItem[]> {
  const params = new URLSearchParams({ profile_id: profileId })
  if (deviceType) params.set('device_type', deviceType)
  if (hourOfDay !== undefined) params.set('hour_of_day', String(hourOfDay))
  return apiFetch<ContinueWatchingItem[]>(`/viewing/continue-watching?${params}`)
}

export function getPausedBookmarks(profileId: string): Promise<ContinueWatchingItem[]> {
  return apiFetch<ContinueWatchingItem[]>(`/viewing/continue-watching/paused?profile_id=${profileId}`)
}

export function dismissBookmark(bookmarkId: string, profileId: string): Promise<BookmarkResponse> {
  return apiFetch<BookmarkResponse>(`/viewing/bookmarks/${bookmarkId}/dismiss?profile_id=${profileId}`, {
    method: 'POST',
  })
}

export function restoreBookmark(bookmarkId: string, profileId: string): Promise<BookmarkResponse> {
  return apiFetch<BookmarkResponse>(`/viewing/bookmarks/${bookmarkId}/restore?profile_id=${profileId}`, {
    method: 'POST',
  })
}

export function getBookmarkByContent(profileId: string, contentId: string): Promise<BookmarkResponse | null> {
  return apiFetch<BookmarkResponse | null>(`/viewing/bookmarks/by-content/${contentId}?profile_id=${profileId}`)
}

export function saveBookmark(profileId: string, payload: BookmarkPayload): Promise<BookmarkResponse> {
  return apiFetch<BookmarkResponse>(`/viewing/bookmarks?profile_id=${profileId}`, {
    method: 'PUT',
    body: JSON.stringify({
      ...payload,
      position_seconds: Math.floor(payload.position_seconds),
      duration_seconds: Math.floor(payload.duration_seconds),
    }),
  })
}

export interface RatingResponse {
  title_id: string
  rating: 1 | -1
}

export function getRating(profileId: string, titleId: string): Promise<RatingResponse | null> {
  return apiFetch<RatingResponse | null>(`/viewing/ratings/${titleId}?profile_id=${profileId}`)
}

export function rateTitle(profileId: string, payload: RatingPayload): Promise<void> {
  return apiFetch<void>(`/viewing/ratings?profile_id=${profileId}`, {
    method: 'POST',
    body: JSON.stringify(payload),
  })
}

export function getWatchlist(profileId: string): Promise<WatchlistItem[]> {
  return apiFetch<WatchlistItem[]>(`/viewing/watchlist?profile_id=${profileId}`)
}

export function addToWatchlist(profileId: string, titleId: string): Promise<void> {
  return apiFetch<void>(`/viewing/watchlist/${titleId}?profile_id=${profileId}`, {
    method: 'POST',
  })
}

export function removeFromWatchlist(profileId: string, titleId: string): Promise<void> {
  return apiFetch<void>(`/viewing/watchlist/${titleId}?profile_id=${profileId}`, {
    method: 'DELETE',
  })
}

File: backend/app/schemas/viewing.py
"""Pydantic schemas for viewing-related endpoints (bookmarks, ratings, watchlist)."""

import uuid
from datetime import datetime

from pydantic import BaseModel, Field, field_validator


# -- Request schemas ----------------------------------------------------------


class BookmarkUpdate(BaseModel):
    """Create or update a playback bookmark."""

    content_type: str = Field(pattern=r"^(movie|episode)$", description="'movie' or 'episode'")
    content_id: uuid.UUID
    position_seconds: int = Field(ge=0)
    duration_seconds: int = Field(ge=1)

    @field_validator("position_seconds", "duration_seconds", mode="before")
    @classmethod
    def truncate_floats(cls, v: object) -> object:
        """Accept float values from clients and truncate to int."""
        if isinstance(v, float):
            return int(v)
        return v


class RatingRequest(BaseModel):
    """Rate a title (thumbs up / thumbs down)."""

    title_id: uuid.UUID
    rating: int = Field(description="Must be -1 (thumbs down) or 1 (thumbs up)")

    def model_post_init(self, __context: object) -> None:
        if self.rating not in (-1, 1):
            raise ValueError("rating must be -1 or 1")


# -- Response schemas ---------------------------------------------------------


class BookmarkResponse(BaseModel):
    """A single bookmark (continue-watching entry)."""

    id: uuid.UUID
    content_type: str
    content_id: uuid.UUID
    position_seconds: int
    duration_seconds: int
    completed: bool
    dismissed_at: datetime | None = None
    updated_at: datetime
    title_info: dict | None = None

    model_config = {"from_attributes": True}


class TitleInfo(BaseModel):
    """Title metadata attached to a Continue Watching item."""

    title: str
    poster_url: str | None = None
    landscape_url: str | None = None
    title_type: str  # 'movie' or 'series'
    age_rating: str | None = None
    episode_title: str | None = None
    season_number: int | None = None
    episode_number: int | None = None


class NextEpisodeInfo(BaseModel):
    """Metadata about the next unwatched episode in a series."""

    episode_id: uuid.UUID
    season_number: int
    episode_number: int
    episode_title: str


class ContinueWatchingItem(BaseModel):
    """A single item in the Continue Watching rail."""

    id: uuid.UUID
    content_type: str
    content_id: uuid.UUID
    position_seconds: int
    duration_seconds: int
    progress_percent: float
    completed: bool = False
    dismissed_at: datetime | None = None
    updated_at: datetime
    resumption_score: float | None = None
    title_info: TitleInfo
    next_episode: NextEpisodeInfo | None = None


class RatingResponse(BaseModel):
    """A user's rating for a title."""

    title_id: uuid.UUID
    rating: int
    created_at: datetime

    model_config = {"from_attributes": True}


class WatchlistItemResponse(BaseModel):
    """A single watchlist entry."""

    title_id: uuid.UUID
    added_at: datetime
    title_info: dict | None = None  # {"title": ..., "poster_url": ...}

    model_config = {"from_attributes": True}
